components_library/boards/lupv_board.py

```diff
@@ -0,0 +1,342 @@
+            cpu.get_mmu().pma_checker = PMAChecker(
+                uncacheable=uncacheable_range
+            )
+    @overrides(AbstractBoard)
+    def has_io_bus(self) -> bool:
+        return True
+    @overrides(AbstractBoard)
+    def get_io_bus(self) -> IOXBar:
+        return self.iobus
+    def has_coherent_io(self) -> bool:
+        return True
+    def get_mem_side_coherent_io_port(self) -> Port:
+        return self.iobus.mem_side_ports
+    @overrides(AbstractBoard)
+    def setup_memory_ranges(self):
+        memory = self.get_memory()
+        mem_size = memory.get_size()
+        self.mem_ranges = [AddrRange(start=0x80000000, size=mem_size)]
+        memory.set_memory_range(self.mem_ranges)
+    def set_workload(
+        self, bootloader: str, disk_image: str, command: Optional[str] = None
+    ):
+        """Setup the full system files
+        See http://resources.gem5.org/resources/riscv-fs for the currently
+        tested kernels and OSes.
+        The command is an optional string to execute once the OS is fully
+        booted, assuming the disk image is setup to run `m5 readfile` after
+        booting.
+        After the workload is set up, this functino will generate the device
+        tree file and output it to the output directory.
+        **Limitations**
+        * Only supports a Linux kernel
+        * Disk must be configured correctly to use the command option
+        * This board doesn't support the command option
+        :param bootloader: The compiled bootloader with the kernel as a payload
+        :param disk_image: A disk image containing the OS data. The first
+            partition should be the root partition.
+        :param command: The command(s) to run with bash once the OS is booted
+        """
+        self.workload.object_file = bootloader
+        image = CowDiskImage(
+            child=RawDiskImage(read_only=True), read_only=False
+        )
+        image.child.image_file = disk_image
+        self.platform.lupio_blk.image = image
+        # Linux boot command flags
+        kernel_cmd = [
+            "earlycon console=ttyLIO0",
+            "root=/dev/lda1",
+            "ro"
+        ]
+        self.workload.command_line = " ".join(kernel_cmd)
+        # Note: This must be called after set_workload because it looks for an
+        # attribute named "disk" and connects
+        self._setup_io_devices()
+        self._setup_pma()
+        # Default DTB address if bbl is built with --with-dts option
+        self.workload.dtb_addr = 0x87E00000
+        # We need to wait to generate the device tree until after the disk is
+        # set up. Now that the disk and workload are set, we can generate the
+        # device tree file.
+        self.generate_device_tree(m5.options.outdir)
+        self.workload.dtb_filename = os.path.join(
+            m5.options.outdir, "device.dtb"
+        )
+    def generate_device_tree(self, outdir: str) -> None:
```

Comment:
Can you add some empty spaces inside your functions to separate the different blocks of code? It's pretty dense and hard to read otherwise.